File: shogun/java/src/us/jubat/example/shogun/Shogun.java
package us.jubat.example.shogun;

import java.io.OutputStreamWriter;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.Random;

import us.jubat.classifier.ClassifierClient;
import us.jubat.classifier.Datum;
import us.jubat.classifier.EstimateResult;
import us.jubat.classifier.TupleStringDatum;
import us.jubat.classifier.TupleStringString;

public class Shogun {
	private final ClassifierClient client;
	private final String name;
	private final Random random;

	public Shogun(ClassifierClient client, String name) {
		this.client = client;
		this.name = name;
		this.random = new Random(0);
	}

	/**
	 * Helper function for making Datum object.
	 * 
	 * @param name
	 * @return
	 */
	private static Datum makeDatum(String name) {
		Datum datum = new Datum();
		datum.num_values = Collections.EMPTY_LIST;
		TupleStringString t = new TupleStringString();
		t.first = "name";
		t.second = name;
		datum.string_values = Arrays.asList(t);
		return datum;
	}

	private static TupleStringDatum makeTrain(String tag, String name) {
		TupleStringDatum pair = new TupleStringDatum();
		pair.first = tag;
		pair.second = makeDatum(name);
		return pair;
	}

	private void train() {
		TupleStringDatum[] trainData = {
				makeTrain("徳川", "家康"),
				makeTrain("徳川", "秀忠"),
				makeTrain("徳川", "家光"),
				makeTrain("徳川", "家綱"),
				makeTrain("徳川", "綱吉"),
				makeTrain("徳川", "家宣"),
				makeTrain("徳川", "家継"),
				makeTrain("徳川", "吉宗"),
				makeTrain("徳川", "家重"),
				makeTrain("徳川", "家治"),
				makeTrain("徳川", "家斉"),
				makeTrain("徳川", "家慶"),
				makeTrain("徳川", "家定"),
				makeTrain("徳川", "家茂"),
				// makeTrain("徳川", "慶喜"),

				makeTrain("足利", "尊氏"), makeTrain("足利", "義詮"),
				makeTrain("足利", "義満"),
				makeTrain("足利", "義持"),
				makeTrain("足利", "義量"),
				makeTrain("足利", "義教"),
				makeTrain("足利", "義勝"),
				makeTrain("足利", "義政"),
				makeTrain("足利", "義尚"),
				makeTrain("足利", "義稙"),
				makeTrain("足利", "義澄"),
				makeTrain("足利", "義稙"),
				makeTrain("足利", "義晴"),
				makeTrain("足利", "義輝"),
				makeTrain("足利", "義栄"),
				// makeTrain("足利", "義昭"),

				makeTrain("北条", "時政"), makeTrain("北条", "義時"),
				makeTrain("北条", "泰時"), makeTrain("北条", "経時"),
				makeTrain("北条", "時頼"), makeTrain("北条", "長時"),
				makeTrain("北条", "政村"), makeTrain("北条", "時宗"),
				makeTrain("北条", "貞時"), makeTrain("北条", "師時"),
				makeTrain("北条", "宗宣"), makeTrain("北条", "煕時"),
				makeTrain("北条", "基時"), makeTrain("北条", "高時"),
				makeTrain("北条", "貞顕"),
		// makeTrain("北条", "守時"),
		};
		// prepare training data
		// predict the last ones (that are commented out)
		List<TupleStringDatum> t = new ArrayList<TupleStringDatum>(
				Arrays.asList(trainData));
		Collections.shuffle(t, random);

		// run train
		client.train(name, t);
	}

	private static EstimateResult findBestResult(List<EstimateResult> res) {
		EstimateResult best = null;
		for (EstimateResult r : res) {
			if (best == null || best.score < r.score) {
				best = r;
			}
		}
		return best;
	}

	private void predict() {
		// predict the last shogun
		Datum[] data = { makeDatum("慶喜"), makeDatum("義昭"), makeDatum("守時"), };
		for (Datum datum : data) {
			List<List<EstimateResult>> res = client.classify(name,
					Arrays.asList(datum));
			// get the predicted shogun name
			System.out.println(findBestResult(res.get(0)).label
					+ datum.string_values.get(0).second);
		}
	}

	public static void main(String[] args) {
		try {
			ClassifierClient client = new ClassifierClient("127.0.0.1", 9199, 1);
			Shogun s = new Shogun(client, "test");
			s.train();
			s.predict();
		} catch (Exception e) {
			e.printStackTrace();
		}
		System.exit(0);
	}
}

File: shogun/cpp/shogun.cpp
#include <iostream>
#include <string>
#include <utility>
#include <vector>

#include <jubatus/client.hpp>

using std::make_pair;
using std::pair;
using std::string;
using std::vector;
using jubatus::client::common::datum;
using jubatus::classifier::estimate_result;
using jubatus::classifier::labeled_datum;

datum make_datum(const string& name) {
	datum d;
	d.add_string("name", name);
	return d;
}

void train(jubatus::classifier::client::classifier client){
	vector<labeled_datum> train_data;
	
	train_data.push_back(labeled_datum("徳川",   make_datum("家康")));
	train_data.push_back(labeled_datum("徳川",   make_datum("秀忠")));
	train_data.push_back(labeled_datum("徳川",   make_datum("家光")));
	train_data.push_back(labeled_datum("徳川",   make_datum("家綱")));
	train_data.push_back(labeled_datum("徳川",   make_datum("綱吉")));
	train_data.push_back(labeled_datum("徳川",   make_datum("家宣")));
	train_data.push_back(labeled_datum("徳川",   make_datum("家継")));
	train_data.push_back(labeled_datum("徳川",   make_datum("吉宗")));
	train_data.push_back(labeled_datum("徳川",   make_datum("家重")));
	train_data.push_back(labeled_datum("徳川",   make_datum("家治")));
	train_data.push_back(labeled_datum("徳川",   make_datum("家斉")));
	train_data.push_back(labeled_datum("徳川",   make_datum("家慶")));
	train_data.push_back(labeled_datum("徳川",   make_datum("家定")));
	train_data.push_back(labeled_datum("徳川",   make_datum("家茂")));
	// train_data.push_back(labeled_datum("徳川",   make_datum("慶喜")));
	
	train_data.push_back(labeled_datum("足利",   make_datum("尊氏")));
	train_data.push_back(labeled_datum("足利",   make_datum("義詮")));
	train_data.push_back(labeled_datum("足利",   make_datum("義満")));
	train_data.push_back(labeled_datum("足利",   make_datum("義持")));
	train_data.push_back(labeled_datum("足利",   make_datum("義量")));
	train_data.push_back(labeled_datum("足利",   make_datum("義教")));
	train_data.push_back(labeled_datum("足利",   make_datum("義勝")));
	train_data.push_back(labeled_datum("足利",   make_datum("義政")));
	train_data.push_back(labeled_datum("足利",   make_datum("義尚")));
	train_data.push_back(labeled_datum("足利",   make_datum("義稙")));
	train_data.push_back(labeled_datum("足利",   make_datum("義澄")));
	train_data.push_back(labeled_datum("足利",   make_datum("義稙")));
	train_data.push_back(labeled_datum("足利",   make_datum("義晴")));
	train_data.push_back(labeled_datum("足利",   make_datum("義輝")));
	train_data.push_back(labeled_datum("足利",   make_datum("義栄")));
	// train_data.push_back(labeled_datum("足利",   make_datum("義昭")));
	
	train_data.push_back(labeled_datum("北条",   make_datum("時政")));
	train_data.push_back(labeled_datum("北条",   make_datum("義時")));
	train_data.push_back(labeled_datum("北条",   make_datum("泰時")));
	train_data.push_back(labeled_datum("北条",   make_datum("経時")));
	train_data.push_back(labeled_datum("北条",   make_datum("時頼")));
	train_data.push_back(labeled_datum("北条",   make_datum("長時")));
	train_data.push_back(labeled_datum("北条",   make_datum("政村")));
	train_data.push_back(labeled_datum("北条",   make_datum("時宗")));
	train_data.push_back(labeled_datum("北条",   make_datum("貞時")));
	train_data.push_back(labeled_datum("北条",   make_datum("師時")));
	train_data.push_back(labeled_datum("北条",   make_datum("宗宣")));
	train_data.push_back(labeled_datum("北条",   make_datum("煕時")));
	train_data.push_back(labeled_datum("北条",   make_datum("基時")));
	train_data.push_back(labeled_datum("北条",   make_datum("高時")));
	train_data.push_back(labeled_datum("北条",   make_datum("貞顕")));
	// train_data.push_back(labeled_datum("北条",   make_datum("守時")));
	
	client.train(train_data);
	
}

void predict(jubatus::classifier::client::classifier client){
	vector<string> test_name;
	test_name.push_back("慶喜");
	test_name.push_back("義昭");
	test_name.push_back("守時");
	
	vector<datum> test_data;
	for(size_t i = 0; i < test_name.size(); ++i){
		test_data.push_back(make_datum(test_name[i]));
	}
	vector<vector<estimate_result> > results = client.classify(test_data);
	
	for (size_t i = 0; i < results.size(); ++i) {
		estimate_result& r = results[i][0];
		for (size_t j = 0; j < results[i].size(); ++j) {
			const estimate_result& temp = results[i][j];
			if(r.score < temp.score){
				r = temp;
			}
		}
		std::string name = test_name[i];
		std::cout << r.label << " " << name <<  std::endl;
	}
}

int main() {
	string host = "127.0.0.1";
	int port = 9199;
	string name = "test";
	
	jubatus::classifier::client::classifier client(host, port, name, 1.0);
	
	train(client);
	predict(client);
	
	return 0;
}